man/vdu.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH VDU "8" "May 2013" "vdu  -- calculates the size of a directory" "System Administration"
.SH NAME
vdu \- calculates the size of a directory
.SH DESCRIPTION
.SS "Usage:"
.IP
vdu \fB\-\-xid\fR <xid> (\fB\-\-space\fR|\-\-inodes) [\-\-blocksize <blocksize>] [\-\-script] <directory>*
.PP
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2006 Enrico Scholz
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/vdu.c
// $Id$    --*- c -*--

// Copyright (C) 2006 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "util.h"
#include <lib/vserver.h>
#include <lib/fmt.h>

#include <stdlib.h>
#include <getopt.h>
#include <stdint.h>
#include <errno.h>
#include <sys/stat.h>
#include <dirent.h>
#include <fcntl.h>

#define ENSC_WRAPPERS_PREFIX	"vdu: "
#define ENSC_WRAPPERS_VSERVER	1
#define ENSC_WRAPPERS_UNISTD	1
#define ENSC_WRAPPERS_DIRENT	1
#define ENSC_WRAPPERS_FCNTL	1
#define ENSC_WRAPPERS_STAT	1
#include <wrappers.h>

#define CMD_HELP		0x1000
#define CMD_VERSION		0x1001
#define CMD_XID			0x2000
#define CMD_SPACE		0x2001
#define CMD_INODES		0x2002
#define CMD_SCRIPT		0x2003
#define CMD_BLOCKSIZE		0x2005

int			wrapper_exit_code = 1;

struct option const
CMDLINE_OPTIONS[] = {
  { "help",           no_argument,       0, CMD_HELP },
  { "version",        no_argument,       0, CMD_VERSION },
  { "xid",            required_argument, 0, CMD_XID },
  { "space",          no_argument,       0, CMD_SPACE },
  { "inodes",         no_argument,       0, CMD_INODES },
  { "script",         no_argument,       0, CMD_SCRIPT },
  { "blocksize",      required_argument, 0, CMD_BLOCKSIZE },
  {0,0,0,0}
};

struct Arguments {
    xid_t		xid;
    bool		space;
    bool		inodes;
    bool		script;
    unsigned long	blocksize;
};

struct Result {
    uint_least64_t	blocks;
    uint_least64_t	inodes;
};

struct TraversalParams {
    struct Arguments const * const	args;
    struct Result * const		result;
};

static void
showHelp(int fd, char const *cmd, int res)
{
  WRITE_MSG(fd, "Usage:\n    ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " --xid <xid> (--space|--inodes) [--blocksize <blocksize>] [--script] <directory>*\n"
	    "\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");

  exit(res);
}

static void
showVersion()
{
  WRITE_MSG(1,
	    "vdu " VERSION " -- calculates the size of a directory\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2006 Enrico Scholz\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

/* basic hash table implementation for inode tracking */
#define HASH_SIZE 103
typedef struct hash_entry {
  struct hash_entry *next;
  ino_t inode;
} hash_entry;

typedef struct hash_table {
  hash_entry *entries[HASH_SIZE];
} hash_table;

static hash_table ht;

static void
hash_init(void)
{
  memset(&ht, 0, sizeof(hash_table));
}

static void
hash_free(void)
{
  int i;
  hash_entry *e, *p;
  for (i = 0; i < HASH_SIZE; i++) {
    for (e = ht.entries[i], p = NULL; e; e = e->next) {
      free(p);
      p = e;
    }
    free(p);
  }
}

static int
hash_insert(ino_t inode)
{
  hash_entry *e, *p;
  unsigned int hashval = inode % HASH_SIZE;

  /* no one else here */
  if (ht.entries[hashval] == NULL) {
    ht.entries[hashval]        = malloc(sizeof(hash_entry));
    ht.entries[hashval]->next  = NULL;
    ht.entries[hashval]->inode = inode;
    return 0;
  }

  for (e = ht.entries[hashval], p = NULL; e; e = e->next) {
    /* already in the hash table */
    if (e->inode == inode)
      return -1;
    else if (e->inode > inode) {
      /* we're first */
      if (p == NULL) {
	ht.entries[hashval]        = malloc(sizeof(hash_entry));
	ht.entries[hashval]->next  = e;
	ht.entries[hashval]->inode = inode;
      }
      /* we're in the middle */
      else {
	p->next        = malloc(sizeof(hash_entry));
	p->next->next  = e;
	p->next->inode = inode;
      }
      return 0;
    }
    p = e;
  }
  /* we're last */
  p->next        = malloc(sizeof(hash_entry));
  p->next->next  = NULL;
  p->next->inode = inode;

  return 0;
}

static void
visitDirEntry(char const *name, dev_t const dir_dev,
	      struct TraversalParams *params);

static void
visitDir(char const *name, struct stat const *expected_stat, struct TraversalParams *params)
{
  int		fd = Eopen(".", O_RDONLY|O_DIRECTORY, 0);
  DIR *		dir;

  EsafeChdir(name, expected_stat);

  dir = Eopendir(".");

  for (;;) {
    struct dirent		*ent = Ereaddir(dir);
    if (ent==0) break;

    if (isDotfile(ent->d_name)) continue;
    visitDirEntry(ent->d_name, expected_stat->st_dev, params);
  }

  Eclosedir(dir);

  Efchdir(fd);
  Eclose(fd);
}

static void
visitDirEntry(char const *name, dev_t const dir_dev,
	      struct TraversalParams *params)
{
  struct stat		st;
  xid_t			xid;
  
  ElstatD(name, &st);

  xid = vc_getfilecontext(name);
  if (xid == params->args->xid &&
      (st.st_nlink == 1 || hash_insert(st.st_ino) != -1)) {
    params->result->blocks += st.st_blocks;
    params->result->inodes += 1;
  }

  if (S_ISDIR(st.st_mode) && dir_dev == st.st_dev)
    visitDir(name, &st, params);
}

static void
visitDirStart(char const *name, struct TraversalParams *params)
{
  struct stat	st;
  int		fd = Eopen(".", O_RDONLY|O_DIRECTORY, 0);

  Estat(name, &st);
  Echdir(name);

  visitDirEntry(".", st.st_dev, params);

  Efchdir(fd);
  Eclose(fd);  
}

int main(int argc, char *argv[])
{
  struct Arguments		args = {
    .xid       = VC_NOCTX,
    .space     = false,
    .inodes    = false,
    .script    = false,
    .blocksize = 1024,
  };
  
  while (1) {
    int		c = getopt_long(argc, argv, "+", CMDLINE_OPTIONS, 0);
    if (c==-1) break;

    switch (c) {
      case CMD_HELP	:  showHelp(1, argv[0], 0);
      case CMD_VERSION	:  showVersion();
      case CMD_XID	:  args.xid = Evc_xidopt2xid(optarg,true); break;
      case CMD_SPACE	:  args.space  = true;                     break;
      case CMD_INODES	:  args.inodes = true;                     break;
      case CMD_SCRIPT	:  args.script = true;                     break;
      case CMD_BLOCKSIZE:
	if (!isNumberUnsigned(optarg, &args.blocksize, false)) {
	  WRITE_MSG(2, "Invalid block size argument: '");
	  WRITE_STR(2, optarg);
	  WRITE_MSG(2, "'; try '--help' for more information\n");
	  return EXIT_FAILURE;
	}
	break;
      default		:
	WRITE_MSG(2, "Try '");
	WRITE_STR(2, argv[0]);
	WRITE_MSG(2, " --help' for more information.\n");
	return 255;
	break;
    }
  }

  if (args.xid==VC_NOCTX)
    WRITE_MSG(2, "No xid specified; try '--help' for more information\n");
  else if (!args.space && !args.inodes)
    WRITE_MSG(2, "Must specify --space or --inodes; try '--help' for more information\n");
  else if (optind==argc)
    WRITE_MSG(2, "No directory specified; try '--help' for more information\n");
  else {
    int		i;
    size_t	len;
    struct Result		result;
    struct TraversalParams	params   = {
      .args   = &args,
      .result = &result
    };

    for (i = optind; i < argc; i++) {
      uint_least64_t		size;
      char			buf[sizeof(size)*3 + 3];
      char const *		delim = "";
      
      result.blocks = 0;
      result.inodes = 0;

      hash_init();
      visitDirStart(argv[i], &params);
      hash_free();

      if (!args.script) {
	WRITE_STR(1, argv[i]);
	WRITE_MSG(1, " ");
      }

      if (args.space) {
	len = utilvserver_fmt_uint64(buf, result.blocks*512 / args.blocksize);
	if (*delim) WRITE_STR(1, delim);
	Vwrite(1, buf, len);
	delim = " ";
      }
      if (args.inodes) {
	len = utilvserver_fmt_uint64(buf, result.inodes);
	if (*delim) WRITE_STR(1, delim);
	Vwrite(1, buf, len);
	delim = " ";
      }	
      WRITE_MSG(1, "\n");
    }
    return EXIT_SUCCESS;
  }

  return EXIT_FAILURE;
}
